### PR TITLE
use fast delete_file_in_range for peer destroy

### PR DESCRIPTION
**Pull request by siddontang**
@ngaut @BusyJay @disksing @zhangjinpeng1987

### CODE REVIEW

**Comment by siddontang** [comment]
Use fast delete_file_in_range to delete data quickly, this function is fast but can't guarantee data consistence.

1. For peer destroy, first set Tombstone and clear raft meta, then use fast delete to delete data and raft log. If we meet panic when fast deleting, the peer is still tombstone when restart and has no side effect. Btw, the dirty data will be cleared up if a newer snapshot for this region applied, and we will introduce compaction filter to handle this dirty data later.  @ngaut , @BusyJay and I discuss that we don't need to introduce another Destroying state, using Tombstone is enough. 
2. For snapshot apply, because we may crash in case 2, so we must force clear the data in snapshot range before applying snapshot. 
3. For log compaction, because we don't care consistent before, so we can use fast delete directly.

**Comment by siddontang** [comment]
Any update? @BusyJay 
If we can't support CompactionFilter now, I think we should use Destroying flag. 
It doesn't conflict with following Compaction work. 
/cc @ngaut

**Comment by BusyJay** [comment]
@ngaut @siddontang PTAL

Instead of introducing a new Destroying flag or using `CompactionFilter`, I add a clean up function to store. It will keep all the valid data ranges, and delete all the invalids (if a key is not in any range of known region, then it's invalid) when startup.

File: src/raftstore/store/engine.rs [hunk 1]
```diff
@@ -240,11 +239,63 @@ pub trait Mutable: Writable {
...
+    for cf in db.cf_names() {
+        try!(delete_in_range_cf(db, cf, start_key, end_key));
+    }
+
+    Ok(())
+}
+
+pub fn delete_in_range_cf(db: &DB, cf: &str, start_key: &[u8], end_key: &[u8]) -> Result<()> {
+    let handle = try!(rocksdb::get_cf_handle(db, cf));
+    try!(db.delete_file_in_range_cf(*handle, start_key, end_key));
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/engine.rs]: If we use this api to destroy region, we'd better increase region size, this api can benefit a lot just when the deleted chunk is large.

Comment (siddontang) [inline src/raftstore/store/engine.rs]: yes, we may use 256mb for region size later.

**Comment by siddontang** [comment]
@BusyJay should update ci script, now we must install rocksdb manually, without brew.

File: src/raftstore/store/peer_storage.rs [hunk 2]
```diff
@@ -428,34 +423,33 @@ impl PeerStorage {
...
-                                  &r.1,
-                                  &mut |key, _| {
-                                      try!(wb.delete(key));
-                                      Ok(true)
-                                  }));
+    /// Delete all meta belong to the region. Results are stored in `wb`.
+    pub fn clear_meta(&self, wb: &WriteBatch) -> Result<()> {
+        let region_id = self.get_region_id();
+        let ranges = [
+            (keys::region_raft_prefix(region_id), keys::region_raft_prefix(region_id + 1)),
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: this may include raft log?

Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: Yes.

**Comment by siddontang** [comment]
LGTM

PTAL @hhkbp2 @zhangjinpeng1987

**Comment by zhangjinpeng1987** [comment]
LGTM

File: src/raftstore/store/engine.rs [hunk 3]
```diff
@@ -240,11 +239,63 @@ pub trait Mutable: Writable {
 impl Mutable for DB {}
 impl Mutable for WriteBatch {}
 
+const MAX_DELETE_KEYS_COUNT: usize = 10000;
```
Comment (hhkbp2) [inline src/raftstore/store/engine.rs]: Why it's 10000?

Comment (BusyJay) [inline src/raftstore/store/engine.rs]: No special reason. It can be any number that is not too small.

**Comment by hhkbp2** [comment]
LGTM

**Pull request merged**